### PR TITLE
[Bugfix] Honor stop strings in beam search

### PR DESCRIPTION
**Pull request by gakkiri**
## Purpose

When `use_beam_search=True`, stop strings from the Chat Completions and Completions APIs were not propagated to `BeamSearchParams`. In addition, the beam-search loops only completed beams on EOS, so a requested stop string was ignored and generation continued until EOS or `max_tokens`.

This PR:

- propagates `stop` from both `ChatCompletionRequest` and `CompletionRequest`;
- adds stop-string support to the shared `BeamSearchParams` contract;
- applies the same stop matching behavior to online and offline beam search;
- reuses the existing stop-string selection semantics, including overlapping
  stop strings and matches spanning token boundaries;
- supports stop strings that end inside a token;
- preserves the triggering token IDs and logprobs while truncating only the
  returned text, matching the regular detokenizer contract;
- preserves the existing offline contract where `BeamSearchSequence.text`
  includes the prompt; and
- keeps the no-stop path free of candidate detokenization overhead.

### Behavior before and after

For example, assume a beam decodes to `"answer STOP trailing text"` and the request specifies `stop=["STOP"]`.

| Behavior | Before this PR | After this PR |
| --- | --- | --- |
| Stop parameter | Dropped while building `BeamSearchParams` | Propagated to online and offline beam search |
| Termination | Beam continues until EOS or `max_tokens` | Beam completes as soon as `STOP` is matched |
| Text with `include_stop_str_in_output=False` | May contain `STOP` and text generated after it | Returns `"answer "` |
| Text with `include_stop_str_in_output=True` | May contain text generated after `STOP` | Returns `"answer STOP"` |
| `finish_reason` | Usually `"length"` or an unrelated EOS stop | `"stop"` |
| `stop_reason` | `None` or an unrelated token ID | `"STOP"` |
| Triggering token IDs and logprobs | Generation continues past the match | Preserved through the token that completes the match |

The same behavior applies when a stop string spans multiple tokens or ends in
the middle of a decoded token. Stop matching only examines generated text, so
a matching string in the prompt does not terminate the beam.

### Duplicate-work check

I searched the open issues and PRs using `beam search stop`, `beam_search stop`, `stop strings beam`, and `include_stop_str_in_output beam`. I did not find an open PR implementing this fix. #36285 also modifies online beam search, but it adds structured-output support and does not address stop strings, so this PR is not a duplicate.

## Test Plan

The regression tests use fake renderers and engine clients. They do not load a model or require a GPU. `--confcutdir` keeps the focused unit-test run isolated from repository-wide model/GPU fixtures.

```bash
uv venv --python 3.12
source .venv/bin/activate
uv pip install -r requirements/cpu.txt pytest pytest-asyncio

.venv/bin/python -m pytest \
  --confcutdir=tests/samplers \
  tests/samplers/test_beam_search_offline.py \
  tests/samplers/test_beam_search_online.py -v
```

The tests cover:

- online and offline stop-string termination;
- stop strings spanning multiple tokens;
- stop strings ending inside a token;
- `include_stop_str_in_output=True` and `False`;
- termination before `max_tokens`;
- preservation of triggering token IDs;
- `finish_reason` and `stop_reason`; and
- propagation from both Chat Completions and Completions requests.

Static checks were also run on all changed Python files:

- `ruff check`
- `ruff format --check`
- the Python 3.12 mypy pre-commit hook
- the typos pre-commit hook
- Python AST parsing
- `git diff --check`

## Test Result

```text
============================== 7 passed in 5.81s ===============================
```

All listed static checks passed.

The original issue was also reproduced before the change and confirmed fixed after the change by the submitter in a GPU environment. The expected observable result is the before/after behavior documented above; model execution and beam scores before the stop match remain unchanged.

### Model evaluation

No model-quality evaluation was run because this change does not modify model execution, logits, beam scores, or candidate token selection. Stop matching is performed after candidate logprobs have already been returned and only changes whether a candidate is moved to the completed-beam set. The affected output contract is covered by the CPU-only behavioral regression tests above.

### Documentation

No documentation update is required. This restores the documented `stop` behavior for beam-search requests without introducing a new API.

### CODE REVIEW

**Comment by gakkiri** [comment]
Hi maintainers, I am a first-time contributor and pre-commit is blocked by the contributor authorization check.

I reproduced the original beam-search stop-string failure and validated this PR end-to-end on a GPU machine; the issue is fully resolved. The focused CPU-only regression tests also pass locally (7 passed), covering online/offline beam search, multi-token and partial-token stop strings, and include/exclude-stop behavior.

Could someone please add the `verified` label to authorize pre-commit?
Thanks!

File: vllm/sampling_params.py [hunk 1]
```diff
@@ -1194,6 +1194,9 @@ class BeamSearchParams(
     length_penalty: float = 1.0
     include_stop_str_in_output: bool = False
     structured_outputs: StructuredOutputsParams | None = None
+    stop: list[str] | None = None
```
Comment (DarkLight1337) [inline vllm/sampling_params.py]: Could you use `StopParam` as the type annotation? Feel free to move the definition into `sampling_params.py`.

Comment (gakkiri) [reply vllm/sampling_params.py]: Yep, done — I moved `StopParam` into `sampling_params.py` and now use it for `BeamSearchParams.stop`. I also moved the string-to-list normalization into `BeamSearchParams.__post_init__` so the protocols can just pass `stop` through directly.

File: vllm/entrypoints/generate/beam_search/online.py [hunk 2]
```diff
@@ -40,6 +44,8 @@ async def beam_search(
         temperature = params.temperature
         length_penalty = params.length_penalty
         include_stop_str_in_output = params.include_stop_str_in_output
+        stop_strings = params.stop
+        assert isinstance(stop_strings, list)
```
Comment (DarkLight1337) [inline vllm/entrypoints/generate/beam_search/online.py]: Could we remove this assertion and handle it properly? For example we could add a pre-computed attribute to `BeamSearchParams` that returns `self.stop` normalized to a list

Comment (gakkiri) [reply vllm/entrypoints/generate/beam_search/online.py]: Ok, I replaced the asserts with a `stop_strings` property on `BeamSearchParams`. I used a regular property since the normalization is cheap, and this way it won't get stale if `stop` changes.

 I also added a couple of tests for multi-beam and encoder-decoder cases.

File: vllm/entrypoints/generate/beam_search/online.py [hunk 3]
```diff
@@ -129,6 +134,7 @@ async def beam_search(
                 if result.outputs[0].logprobs is not None:
                     logprobs = result.outputs[0].logprobs[0]
                     for token_id, logprob_obj in logprobs.items():
+                        token_id = int(token_id)
```
Comment (DarkLight1337) [inline vllm/entrypoints/generate/beam_search/online.py]: Why is this needed?

File: vllm/entrypoints/generate/beam_search/online.py [hunk 4]
```diff
@@ -192,6 +220,8 @@ async def beam_search(
         best_beams = sorted_completed[:beam_width]
 
         for beam in best_beams:
+            if beam.text is not None:
```
Comment (DarkLight1337) [inline vllm/entrypoints/generate/beam_search/online.py:224]: Please add a comment explaining which cases this corresponds to

File: vllm/entrypoints/generate/beam_search/online.py (hunk 3 above)
Comment (gakkiri) [reply vllm/entrypoints/generate/beam_search/online.py]: This cast was already in the original code here:
[https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/generate/beam_search/online.py#L152](https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/generate/beam_search/online.py#L152)

I moved it up because the new stop-string path uses `token_id` before it gets added to `candidates`, so both paths would use a plain Python `int`.

That said, the logprob keys are already typed as `int`, so I'll move the cast back to its original place to keep the diff smaller.

File: vllm/entrypoints/generate/beam_search/online.py (hunk 4 above)
Comment (gakkiri) [reply vllm/entrypoints/generate/beam_search/online.py:224]: This handles stop strings that end inside a token. For example, if a token decodes to `" organization"` and the stop string is `"gani"`, the returned text should end in `"or"` (or `"organi"` when the stop string is included), while the full token ID is still retained for token IDs and logprobs.

Re-decoding that token here would bring back the stop string or the trailing `"zation"`.

I've added a comment there to clarify this case.

**Comment by DarkLight1337** [comment]
@njhill please check if this is how stop strings is supposed to work in general
